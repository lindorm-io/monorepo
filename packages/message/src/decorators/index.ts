export * from "./CorrelationField";
export * from "./DelayField";
export * from "./Field";
export * from "./Generated";
export * from "./IdentifierField";
export * from "./MandatoryField";
export * from "./Message";
export * from "./OnConsume";
export * from "./OnCreate";
export * from "./OnPublish";
export * from "./OnValidate";
export * from "./PersistentField";
export * from "./Priority";
export * from "./PriorityField";
export * from "./Schema";
export * from "./TimestampField";
export * from "./Topic";
