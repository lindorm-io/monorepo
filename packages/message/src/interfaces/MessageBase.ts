export interface IMessageBase {
  id: string;
  timestamp: Date;
}
