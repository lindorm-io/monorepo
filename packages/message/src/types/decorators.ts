import { ZodType } from "zod";
import { MetaFieldFallback, MetaFieldType, MetaGeneratedStrategy } from "./metadata";

export type FieldDecoratorOptions = {
  enum?: any;
  fallback?: MetaFieldFallback;
  max?: number;
  min?: number;
  nullable?: boolean;
  optional?: boolean;
  schema?: ZodType;
  type?: MetaFieldType;
};

export type MessageDecoratorOptions = {
  name?: string;
  namespace?: string;
  topic?: string;
};

export type GeneratedDecoratorOptions = {
  strategy?: MetaGeneratedStrategy;
  length?: number;
  max?: number;
  min?: number;
};

export type HookDecoratorCallback = (message: any) => void;
