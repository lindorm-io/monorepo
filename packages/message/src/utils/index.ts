export * from "./default-create-message";
export * from "./default-generate-message";
export * from "./default-validate-message";
export * from "./get-topic-name";
export * from "./global";
