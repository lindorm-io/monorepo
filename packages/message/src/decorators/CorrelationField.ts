import { randomUUID } from "crypto";
import { globalMessageMetadata } from "../utils";

export function CorrelationField(): PropertyDecorator {
  return function (target, key) {
    globalMessageMetadata.addField({
      target: target.constructor,
      key: key.toString(),
      decorator: "CorrelationField",
      enum: null,
      fallback: () => randomUUID(),
      max: null,
      min: null,
      nullable: false,
      optional: false,
      schema: null,
      type: "uuid",
    });
  };
}
