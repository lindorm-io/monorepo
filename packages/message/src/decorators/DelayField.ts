import { globalMessageMetadata } from "../utils";

export function DelayField(): PropertyDecorator {
  return function (target, key) {
    globalMessageMetadata.addField({
      target: target.constructor,
      key: key.toString(),
      decorator: "DelayField",
      enum: null,
      fallback: 0,
      max: null,
      min: 0,
      nullable: false,
      optional: false,
      schema: null,
      type: "integer",
    });
  };
}
