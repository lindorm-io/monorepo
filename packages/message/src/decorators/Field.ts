import { isObject, isString } from "@lindorm/is";
import { FieldDecoratorOptions, MetaFieldType } from "../types";
import { globalMessageMetadata } from "../utils";

export function Field(
  type?: MetaFieldType,
  options?: Omit<FieldDecoratorOptions, "type">,
): PropertyDecorator;
export function Field(options?: FieldDecoratorOptions): PropertyDecorator;
export function Field(arg1?: any, arg2?: any): PropertyDecorator {
  return function (target, key) {
    const type = isString<MetaFieldType>(arg1)
      ? arg1
      : isObject<FieldDecoratorOptions>(arg1)
        ? arg1.type
        : null;

    const options: FieldDecoratorOptions | undefined = isObject(arg1) ? arg1 : arg2;

    globalMessageMetadata.addField({
      target: target.constructor,
      key: key.toString(),
      decorator: "Field",
      enum: options?.enum ?? null,
      fallback: options?.fallback ?? null,
      max: options?.max ?? null,
      min: options?.min ?? null,
      nullable: options?.nullable ?? false,
      optional: options?.optional ?? false,
      schema: options?.schema ?? null,
      type: type ?? null,
    });
  };
}
