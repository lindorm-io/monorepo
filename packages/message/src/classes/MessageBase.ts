import { IdentifierField, TimestampField } from "../decorators";
import { IMessageBase } from "../interfaces";

export abstract class MessageBase implements IMessageBase {
  @IdentifierField()
  public readonly id!: string;

  @TimestampField()
  public readonly timestamp!: Date;
}
