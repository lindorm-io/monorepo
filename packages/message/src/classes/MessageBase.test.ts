import { Message } from "../decorators";
import { globalMessageMetadata } from "../utils";
import { MessageBase } from "./MessageBase";

describe("MessageBase", () => {
  test("should have expected properties", () => {
    @Message()
    class TestMessageBase extends MessageBase {}

    expect(new TestMessageBase()).toMatchSnapshot();
    expect(globalMessageMetadata.get(TestMessageBase)).toMatchSnapshot();
  });
});
