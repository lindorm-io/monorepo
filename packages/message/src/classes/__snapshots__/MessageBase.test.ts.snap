// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageBase should have expected properties 1`] = `
TestMessageBase {
  "id": undefined,
  "timestamp": undefined,
}
`;

exports[`MessageBase should have expected properties 2`] = `
{
  "fields": [
    {
      "decorator": "IdentifierField",
      "enum": null,
      "fallback": [Function],
      "key": "id",
      "max": null,
      "min": null,
      "nullable": false,
      "optional": false,
      "schema": null,
      "type": "uuid",
    },
    {
      "decorator": "TimestampField",
      "enum": null,
      "fallback": [Function],
      "key": "timestamp",
      "max": null,
      "min": null,
      "nullable": false,
      "optional": false,
      "schema": null,
      "type": "date",
    },
  ],
  "generated": [],
  "hooks": [],
  "message": {
    "decorator": "Message",
    "name": null,
    "namespace": null,
    "topic": null,
  },
  "schemas": [],
}
`;
